src/rgen/Microsoft.Macios.Generator/TabbedStringBuilder.cs

```diff
@@ -0,0 +1,188 @@
+/// <example>
+/// var classBlock = new TabbedStringBuilder (sb);
+/// classBlock.AppendFormatLine ("public static NSString? GetConstant (this {0} self)", enumSymbol.Name);
+/// // open a new {} block, no need to keep track of the indentation, the new block has it
+/// using (var getConstantBlock = classBlock.CreateBlock (isBlock: true)) {
+///    // write the contents of the method here.
+/// }
+/// </example>
+/// </summary>
+class TabbedStringBuilder : IDisposable {
+	readonly StringBuilder sb;
+	readonly uint tabCount;
+	readonly string tabs;
+	readonly bool isBlock;
+	bool disposed;
+
+	/// <summary>
+	/// Created a new tabbed string builder that will use the given sb to write code.
+	/// </summary>
+	/// <param name="builder">The string builder to be used to write code.</param>
+	/// <param name="currentCount">The original tab size.</param>
+	/// <param name="block">States if we are creating a {} block.</param>
+	public TabbedStringBuilder (StringBuilder builder, uint currentCount = 0, bool block = false)
+	{
+		sb = builder;
+		isBlock = block;
+		if (isBlock) {
+			// increase by 1 because we are in a block
+			tabCount = currentCount + 1;
+			var braceTab = new string ('\t', (int) (tabCount - 1));
+			this.sb.AppendLine ($"{braceTab}{{");
+		} else {
+			tabCount = currentCount;
+		}
+		tabs = new string ('\t', (int) tabCount);
+	}
+
+	/// <summary>
+	/// Append a new empty line to the string builder using the correct tab size.
+	/// </summary>
+	/// <returns>The current tabbed string builder.</returns>
+	public TabbedStringBuilder AppendLine ()
+	{
+		sb.AppendLine ();
+		return this;
+	}
+
+	/// <summary>
+	/// Append a line, but do not add a \n
+	/// </summary>
+	/// <param name="line"></param>
+	/// <returns></returns>
+	public TabbedStringBuilder Append (string line)
+	{
+		if (string.IsNullOrWhiteSpace (line)) {
+			sb.Append (line);
+		} else {
+			sb.Append ($"{tabs}{line}");
+		}
+		return this;
+	}
+
+	/// <summary>
+	/// Append a new tabbed line.
+	/// </summary>
+	/// <param name="line">The line to append.</param>
+	/// <returns>The current builder.</returns>
+	public TabbedStringBuilder AppendLine (string line)
+	{
+		if (string.IsNullOrWhiteSpace (line)) {
+			sb.AppendLine (line);
+		} else {
+			sb.AppendLine ($"{tabs}{line}");
+		}
+		return this;
+	}
+
+	public TabbedStringBuilder AppendFormatLine (string format, params object [] args)
+	{
+		sb.AppendFormat ($"{tabs}{format}", args);
+		sb.AppendLine ();
+		return this;
+	}
+
+	/// <summary>
+	/// Append a new raw literal by prepending the correct indentation.
+	/// </summary>
+	/// <param name="rawString">The raw string to append.</param>
+	/// <returns>The current builder.</returns>
+	public TabbedStringBuilder AppendRaw (string rawString)
+	{
+		// we will split the raw string in lines and then append them so that the
+		// tabbing is correct
+		var lines = rawString.Split (['\n'], StringSplitOptions.None);
```

Comment:
It would be nice to implement this in a way that doesn't allocate temporary arrays.

.NET 9 has a new feature where you can split a string and enumerate over the results without allocating memory, it's described somewhere around here: https://devblogs.microsoft.com/dotnet/performance-improvements-in-net-9/#span-span-and-more-span

https://learn.microsoft.com/en-us/dotnet/api/system.memoryextensions.split?view=net-8.0